pkg/operator/operator_config.go

```diff
@@ -296,242 +293,64 @@ func (r *operatorConfigReconciler) ensureRuleEvaluatorSecrets(ctx context.Contex
-							},
-							{
-								Name:      configOutVolumeName,
-								MountPath: configOutDir,
-							},
-							{
-								Name:      rulesVolumeName,
-								MountPath: rulesDir,
-								ReadOnly:  true,
-							},
-							{
-								Name:      secretVolumeName,
-								MountPath: secretsDir,
-								ReadOnly:  true,
-							},
-						},
-						Resources: corev1.ResourceRequirements{
-							Requests: corev1.ResourceList{
-								corev1.ResourceCPU:    *resource.NewScaledQuantity(5, resource.Milli),
-								corev1.ResourceMemory: *resource.NewScaledQuantity(16, resource.Mega),
-							},
-							// Set no limit on CPU as it's a throttled resource.
-							Limits: corev1.ResourceList{
-								corev1.ResourceMemory: *resource.NewScaledQuantity(32, resource.Mega),
-							},
-						},
-						SecurityContext: minimalSecurityContext(),
-					},
-				},
-				Volumes: []corev1.Volume{
-					{
-						// Rule-evaluator input Prometheus config.
-						Name: configVolumeName,
-						VolumeSource: corev1.VolumeSource{
-							ConfigMap: &corev1.ConfigMapVolumeSource{
-								LocalObjectReference: corev1.LocalObjectReference{
-									Name: NameRuleEvaluator,
-								},
-							},
-						},
-					}, {
-						// Generated rule-evaluator output Prometheus config.
-						Name: configOutVolumeName,
-						VolumeSource: corev1.VolumeSource{
-							EmptyDir: &corev1.EmptyDirVolumeSource{},
-						},
-					}, {
-						// Generated rules yaml files via the "rules" runtime controller.
-						Name: rulesVolumeName,
-						VolumeSource: corev1.VolumeSource{
-							ConfigMap: &corev1.ConfigMapVolumeSource{
-								LocalObjectReference: corev1.LocalObjectReference{
-									Name: nameRulesGenerated,
-								},
-							},
-						},
-					}, {
-						// Mirrored config secrets (config specified as filepaths).
-						Name: secretVolumeName,
-						VolumeSource: corev1.VolumeSource{
-							Secret: &corev1.SecretVolumeSource{
-								SecretName: RulesSecretName,
-							},
-						},
-					},
-				},
-				// Collector service account used for K8s endpoints-based SD.
-				// TODO(pintohutch): confirm minimum serviceAccount credentials needed for rule-evaluator
-				// and create dedicated serviceAccount.
-				ServiceAccountName:           NameCollector,
-				AutomountServiceAccountToken: ptr(true),
-				PriorityClassName:            r.opts.PriorityClass,
-				HostNetwork:                  r.opts.HostNetwork,
-			},
-		},
-	}
-	return &appsv1.Deployment{
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: r.opts.OperatorNamespace,
-			Name:      NameRuleEvaluator,
-		},
-		Spec: deploy,
+		deploy.Spec.Template.Spec.Containers[i].Env = repl
 	}
+
+	// Upsert rule-evaluator Deployment.
+	// We've to use Patch() as Update() with update we'll get stuck in an infinite loop as each
+	// update increases the generation since we override the managed fields metadata, which is
+	// set by the kube-controller-manager, which wants to own an annotation on deployments.
```

Comment:
Update changed which field in particular? `observedGeneration` (owned by `kube-controller-manager`). `.metadata.generation` doesn't show up as a `managedFields` in the YAML expansion.

The only thing we're updating here is the `.spec.containers[*].env` field, which the `kube-controller-manager` doesn't own...

Reply:
Right the issue may no longer be relevant as we're only setting a small set of fields. I just didn't want to change too many things at once honestly.